### PR TITLE
[WIP] Add starboard functionality

### PR DESCRIPTION
**Pull request by calledude**
Putting a draft up for feedback. This is what has been done so far to migrate starboard responsibilities to MODiX. 

So far the basic groundwork has been laid. Starred messages above a certain threshold will be put on the designated starboard channel with a URL linking to the context in which the quoted message was sent.

#### TODO: 
* func ret bool => Starred message exists in designated starboard channel
* func void => Delete starred message in designated starboard channel
* func void => Update starred message in designated starboard channel
* Various logic to implement around the above.

### CODE REVIEW

File: Modix.Services/Starboard/StarboardService.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+        //TODO: Function<bool>: => Starred message exists in designated starboard channel
+
+        //TODO: Function: => Delete starred message in designated starboard channel
+
+        //TODO: Function: => Update starred message in designated starboard channel
+
+
+        public Task<bool> IsStarReaction(SocketReaction reaction)
+            => Task.FromResult(reaction.Emote.Name == ReactionEmote);
```
[the remark this answers is not in the document]
Comment (calledude) [reply Modix.Services/Starboard/StarboardService.cs]: There isn't, you are absolutely correct :)
I blame brain overload while getting used to the codebase ;)

File: Modix.Services/Starboard/StarboardService.cs [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+        public Task<int> GetReactionCount(IUserMessage message, ReactionAdded reaction)
+            => Task.FromResult(message.Reactions[reaction.Reaction.Emote].ReactionCount);
+
+        public async Task<bool> IsAboveReactionThreshold(IUserMessage message, ReactionAdded reaction)
+        {
+            return await GetReactionCount(message, reaction) >= 2;
```
[the remark this answers is not in the document]
Comment (calledude) [reply Modix.Services/Starboard/StarboardService.cs]: Probably *won't* be. I'm prioritizing the general functionality for now. But it is indeed something to look into further on.

**Comment by calledude** [comment]
Basic functionality added. 
#### It currently does as following: 
* Deletes messages on starboard when they fall below the threshold. 
* Checks if the message already exists on the designated starboard-channel. 
* Updates messages with new reaction-data.
* Adds a starred message in embed-form to the starboard when the message gets enough stars (currently hardcoded to `>= 2`) 

Currently, only one starboard per guild works.

#### TODO: 
* Figure out where to put clickable link to original message. Preferably in markdown format --> [Show More](link)
* Unhardcode thresholds?
* Add a `Dictionary<Emoji, int>` to compare against instead.

File: Modix.Data/Models/Core/DesignatedChannelType.cs [hunk 3]
```diff
@@ -24,6 +24,10 @@ public enum DesignatedChannelType
         /// <summary>
         /// Defines a channel that is not subject to auto-moderation behaviors of the moderation feature.
         /// </summary>
-        Unmoderated
+        Unmoderated,
+        /// <summary>
+        /// Defines a channel to which starred messages are sent.
+        /// </summary>
+        Starboard
```
Comment (Scott-Caldwell) [inline Modix.Data/Models/Core/DesignatedChannelType.cs]: There's a corresponding enum for the web interface as well that needs to be edited.
https://github.com/discord-csharp/MODiX/blob/master/Modix/ClientApp/src/models/moderation/ChannelDesignation.ts

File: Modix.Services/Starboard/StarboardService.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
+﻿using System.Linq;
+using System.Threading.Tasks;
+using Discord;
+using Discord.WebSocket;
+using Modix.Services.Core;
+using Modix.Data.Models.Core;
+
+namespace Modix.Services.Starboard
+{
+    public class StarboardService
```
Comment (Scott-Caldwell) [inline Modix.Services/Starboard/StarboardService.cs]: Might be nice to have a corresponding interface like the other services have, and some documentation comments for the public methods.

File: Modix.Services/Starboard/StarboardHandler.cs [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+
+        //TODO: Figure out where to put clickable link to original message
+        // Preferably in markdown format --> [Show More](link)
+        public Embed GetStarEmbed(IGuildChannel channel, IUserMessage message)
+        {
+            var author = message.Author as IGuildUser;
+            var builder = new EmbedBuilder()
+                .WithTimestamp(message.Timestamp)
+                .WithColor(new Color(255, 234, 174))
+                .WithAuthor(author.Nickname ?? author.Username, author.GetAvatarUrl())
```
Comment (Scott-Caldwell) [inline Modix.Services/Starboard/StarboardHandler.cs]: If you do `author.GetAvatarUrl() ?? author.GetDefaultAvatarUrl()`, I think it should handle cases where the user just has the default Discord logo avatar

**Pull request merged**